src/vs/workbench/contrib/chat/common/chatModel.ts

```diff
@@ -82,54 +89,86 @@ export class ChatRequestModel implements IChatRequestModel {
-		this._responseRepr = value;
-		this._responseParts = [{ string: value }];
+	constructor(value: IMarkdownString | (IMarkdownString | IChatResponseProgressFileTreeData)[]) {
+		this._responseData = Array.isArray(value) ? value : [value];
+		this._responseParts = Array.isArray(value) ? value.map((v) => ('value' in v ? { string: v } : { treeData: v })) : [{ string: value }];
+		this._responseRepr = this._responseParts.map((part) => {
+			if ('treeData' in part) {
+				return '';
+			}
+			return part.string.value;
+		}).join('\n');
+	}
+
+	asString(): string {
+		return this._responseRepr;
 	}
 
-	updateContent(responsePart: string | { placeholder: string; resolvedContent?: Promise<string> }, quiet?: boolean): void {
+	updateContent(responsePart: string | { placeholder: string; resolvedContent?: Promise<string | { treeData: IChatResponseProgressFileTreeData }> }, quiet?: boolean): void {
 		if (typeof responsePart === 'string') {
 			const responsePartLength = this._responseParts.length - 1;
 			const lastResponsePart = this._responseParts[responsePartLength];
 
-			if (lastResponsePart.resolving === true) {
-				// The last part is resolving, start a new part
+			if (lastResponsePart.resolving === true || 'treeData' in lastResponsePart) {
+				// The last part is resolving or a tree data item, start a new part
 				this._responseParts.push({ string: new MarkdownString(responsePart) });
 			} else {
-				// Combine this part with the last, non-resolving part
+				// Combine this part with the last, non-resolving string part
 				this._responseParts[responsePartLength] = { string: new MarkdownString(lastResponsePart.string.value + responsePart) };
 			}
 
 			this._updateRepr(quiet);
-		} else {
+		} else if ('placeholder' in responsePart) {
 			// Add a new resolving part
 			const responsePosition = this._responseParts.push({ string: new MarkdownString(responsePart.placeholder), resolving: true }) - 1;
 			this._updateRepr(quiet);
 
 			responsePart.resolvedContent?.then((content) => {
 				// Replace the resolving part's content with the resolved response
-				this._responseParts[responsePosition] = { string: new MarkdownString(content), resolving: true };
-				this._updateRepr(quiet);
+				if (typeof content === 'string') {
+					this._responseParts[responsePosition] = { string: new MarkdownString(content), resolving: true };
+					this._updateRepr(quiet);
+				} else if (content.treeData) {
+					this._responseParts[responsePosition] = { treeData: content.treeData, resolving: true };
+					this._updateRepr(quiet);
+				}
 			});
 		}
 	}
 
 	private _updateRepr(quiet?: boolean) {
-		this._responseRepr = new MarkdownString(this._responseParts.map(r => r.string.value).join('\n\n'));
+		this._responseData = this._responseParts.map(part => {
+			if ('treeData' in part) {
+				return part.treeData;
+			}
+			return part.string;
+		});
+
+		this._responseRepr = this._responseParts.map(part => {
+			if ('treeData' in part) {
+				return '';
```

Comment:
I suppose we'll need something for this, for accessibility, at some point, you could ask Megan about that. Should the accessibility view show the text representation of the tree?

Reply:
Yeah, I want to tackle that separately, I think ideally we would be able to tab to the tree and then it would read whatever the explorer currently reads

Reply:
It will also need to show up in the accessible view, and when you just focus the tree row